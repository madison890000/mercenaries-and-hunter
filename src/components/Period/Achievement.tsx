import Tag from '../Tag';
import React from 'react';

interface AchievementProps {
    title: string;
    categories: string[];
}

const Achievement = ({ title, categories }: AchievementProps) => (
    <li>
        {title}
        <span>
            {categories?.map(c => (
                <Tag>{c}</Tag>
            ))}
        </span>
    </li>
);

export default Achievement;
