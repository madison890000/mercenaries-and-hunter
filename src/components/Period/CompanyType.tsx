import Tag from '../Tag';
import React from 'react';
import DataModel from '../../models/types';

interface CompanyTypeProps {
    companyType: DataModel.CompanyType;
}

const CompanyType = ({ companyType }: CompanyTypeProps) => (
    <span>
        {companyType === DataModel.CompanyType.Enterprise && <Tag type="less">Enterprise</Tag>}
        {companyType === DataModel.CompanyType.Startup && <Tag type="less">Startup</Tag>}
    </span>
);

export default CompanyType;
