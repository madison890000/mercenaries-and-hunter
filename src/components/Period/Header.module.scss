.header {
    font-weight: bold;

    & > span {
        margin-right: var(--inline-block-margin);
    }

    @media screen and (max-width: 414px) {
        width: 300px;

        & > span {
            display: block;
        }
    }
}

.job {
    & > span {
        margin-right: var(--inline-block-margin);
    }
}

.companyName {
    font-weight: normal;
    color: var(--period-sub-title-color);
}
