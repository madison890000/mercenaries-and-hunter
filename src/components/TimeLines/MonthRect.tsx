import React from 'react';

const MonthRect = ({ gap, color, index, width }: { index: number; width: number; color?: string; gap: number }) => {
    return <rect width={width} height="10" x={index * (width + gap)} y="5" fill={color} />;
};

export default MonthRect;
