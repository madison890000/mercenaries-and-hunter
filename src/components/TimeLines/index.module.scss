.timelineContainer {
  width: 100%;
  overflow: hidden;
  position: relative;
}

.yearContainer {
  display: flex;
}

.month-rect {
  width: 10px;
  height: 10px;
}

.jobName {
  font-size: 12px;
  width: 80px;
}
