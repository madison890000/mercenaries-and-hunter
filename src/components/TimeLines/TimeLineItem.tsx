import React, { useMemo } from 'react';
import { getMonthCountFromStartAndEnd } from '../../utils/date';
import MonthRect from './MonthRect';
import { getScreenDevice, ScreenDevice } from '../../utils/device';
import { BASE_MONTH_SVG_HEIGHT, BASE_RECT_WIDTH, RECT_GAP } from '../../constants/widths';

export const getBaseRectFull = (device: ScreenDevice) => {
    switch (device) {
        case ScreenDevice.A4:
            return BASE_RECT_WIDTH - 2;
        case ScreenDevice.Mobile:
            return BASE_RECT_WIDTH - 4;
        case ScreenDevice.PC:
            return BASE_RECT_WIDTH;
        default:
            return BASE_RECT_WIDTH;
    }
};

const TimeLineItem = ({ start, end, periodColor }: { start: Date; end?: Date; periodColor: string }) => {
    const { rectFullWidth, totalRects, rects, rectWidth } = useMemo(() => {
        const totalRects = getMonthCountFromStartAndEnd(start, end ?? new Date()) + 1;
        const rectFullWidth = getBaseRectFull(getScreenDevice());
        const rectWidth = rectFullWidth - RECT_GAP;
        let rects = new Array(totalRects).fill(0);
        return {
            rectFullWidth,
            totalRects,
            rects,
            rectWidth
        };
    }, [start, end]);
    return (
        <svg width={rectFullWidth * totalRects} height={BASE_MONTH_SVG_HEIGHT}>
            <g>
                {rects?.map((e, index) => (
                    <MonthRect key={index} width={rectWidth} gap={RECT_GAP} color={periodColor} index={index} />
                ))}
            </g>
        </svg>
    );
};

export default TimeLineItem;
