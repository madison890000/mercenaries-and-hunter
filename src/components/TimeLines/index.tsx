import React from 'react';
import styles from './index.module.scss';
import Period from '../../lib/Period';
import { findPeriodByDate, getMonthCountInStartAndEnd } from '../../utils/date';
import MonthRect from './MonthRect';
import { isMobilePhone } from '../../utils/device';


const PAGE_PADDING = 24;

const TimeLines = ({ periods, periodColors }: { periods: Period[]; periodColors: { [key: string]: string } }) => {
    const start = periods?.[0]?.start;
    const end = periods?.[periods?.length - 1]?.end;
    const width = isMobilePhone() ? window.innerWidth - 24 : window.innerWidth - PAGE_PADDING - 20;
    const rectGap = 1;
    const totalRects = getMonthCountInStartAndEnd(start, end ?? new Date()) + 1;
    const rectFullWidth = width / totalRects;
    const rectWidth = rectFullWidth - rectGap;
    let rectArray = new Array(totalRects).fill(0)
        ?.map((_, index) => {
            const year = start.getFullYear() + Math.floor((start.getMonth() + index) / 12);
            const month = (start.getMonth() + index) % 12;
            const period = findPeriodByDate(new Date(Number(year), month, 1), periods)
            return {
                index: index,
                period: period,
                color: period ? periodColors[period?.id] : '',
            }
        })
    return (
        <div className={styles.timelineContainer}>
            <div>
                <svg width={width} height="20" style={{ backgroundColor: 'E0E0E0' }}>
                    <g>
                        {rectArray?.map((e, index) => (
                            <MonthRect
                                width={rectWidth}
                                gap={rectGap}
                                color={e?.color}
                                index={e?.index}
                            />
                        ))}
                    </g>
                </svg>
            </div>
            <div style={{
                height: 48
            }}>
                {periods?.map((p, index) => {
                    const translateX = (getMonthCountInStartAndEnd(start, p?.start) + 1) * rectFullWidth;
                    return (
                        <span
                            style={{
                                position: 'absolute',
                                transform: `translateX(${translateX}px)`
                            }}
                            className={styles.jobName}
                        >
                        {p?.jobPositionLevel} {p?.jobPosition}
                    </span>
                    )
                })}
            </div>
        </div>
    )
}

export default TimeLines
