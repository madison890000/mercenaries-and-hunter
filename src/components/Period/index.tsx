import React from 'react';
import styles from './index.module.scss';
import DataModel from '../../models/types';
import Times from '../Times';
import Achievement from './Achievement';
import TimeLineItem from '../TimeLines/TimeLineItem';
import { defineMessages, useIntl } from 'react-intl';
import Header from './Header';
import StringWithID from '../../models/StringWithID';
import capitalize from '../../utils/capitalize';

interface PeriodProps {
    start: Date;
    periodColor: string;
    end?: Date;
    keywords: StringWithID[];
    companyName: string;
    companyType: DataModel.CompanyType;
    jobPositionLevel: DataModel.JobPositionLevel;
    jobPosition: DataModel.JobPosition;
    achievements: DataModel.Achievement[];
    jobSummaries: StringWithID[];
}

const messages = defineMessages({
    jobSummaries: {
        id: 'component.period.jobSummaries',
        defaultMessage: 'Job summaries'
    },
    achievements: {
        id: 'component.period.achievements',
        defaultMessage: 'Achievements'
    }
});
const Period = ({
    start,
    periodColor,
    end,
    companyName,
    companyType,
    keywords,
    jobPositionLevel,
    jobPosition,
    achievements,
    jobSummaries
}: PeriodProps) => {
    const intl = useIntl();
    return (
        <div className={styles.periodContainer}>
            <div className={styles.period}>
                <div>
                    <Header
                        jobPosition={jobPosition}
                        companyName={companyName}
                        companyType={companyType}
                        keywords={keywords}
                        jobPositionLevel={jobPositionLevel}
                    />
                </div>
                <div className={styles.timeline}>
                    <TimeLineItem start={start} end={end} periodColor={periodColor} />
                    <Times start={start} end={end} />
                </div>
            </div>

            <div>
                <h5>{intl.formatMessage(messages.jobSummaries)}:</h5>
                <ul>
                    {jobSummaries?.map(summary => (
                        <li key={summary.id}>{capitalize(summary.toString())}</li>
                    ))}
                </ul>
            </div>
            <div>
                <h5>{intl.formatMessage(messages.achievements)}:</h5>
                <ul className={styles.achievements}>
                    {achievements?.map(achievement => (
                        <Achievement
                            key={achievement.id}
                            title={achievement?.text}
                            categories={achievement?.categories}
                        />
                    ))}
                </ul>
            </div>
        </div>
    );
};

export default Period;
