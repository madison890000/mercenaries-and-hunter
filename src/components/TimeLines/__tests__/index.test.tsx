import React from 'react';
import TimeLines from '../index';
import renderer from 'react-test-renderer';
import DataModel from '../../../Model/types';
import Company from '../../../Model/Company';
import Period from '../../../Model/Period';

let windowSpy: any;

beforeEach(() => {
    windowSpy = jest.spyOn(window, 'window', 'get');
});

afterEach(() => {
    windowSpy.mockRestore();
});
describe('renders TimeLine ', () => {
    const testCompanyName = 'test company name';
    const testCompany = new Company({
        name: testCompanyName
    });
    const start = new Date('2012-01-01');
    const end = new Date('2013-01-01');
    const keywords = ['React'] as DataModel.Keyword[];
    const jobPositionLevel = DataModel.JobPositionLevel.Senior;
    const jobPosition = 'Software Engineer';
    const achievements = [] as DataModel.Achievement[];
    const jobSummaries = [] as string[];
    const testPeriod = new Period({
        solutionsOfHowToImplement: [],
        company: testCompany,
        jobSummaries,
        jobPositionLevel,
        jobPosition,
        achievements,
        start,
        end,
        keywords
    });
    test('TimeLine one period', () => {
        const TimeLineOne = renderer
            .create(
                <TimeLines
                    periods={[testPeriod]}
                    periodColors={{
                        [testPeriod.id]: '#FF6666'
                    }}
                    width={1000}
                />
            )
            .toJSON();
        expect(TimeLineOne).toMatchSnapshot();
    });
});
