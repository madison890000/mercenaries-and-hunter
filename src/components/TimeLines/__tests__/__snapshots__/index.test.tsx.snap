// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TimeLine  TimeLine one period 1`] = `
<div
  className="timelineContainer"
  style={
    Object {
      "width": "1000px",
    }
  }
>
  <div>
    <svg
      height={20}
      style={
        Object {
          "backgroundColor": "#E0E0E0",
        }
      }
      width={1000}
    >
      <g>
        <rect
          fill="#FF6666"
          height="10"
          width={75.92307692307692}
          x={0}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={75.92307692307692}
          x={76.92307692307692}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={75.92307692307692}
          x={153.84615384615384}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={75.92307692307692}
          x={230.76923076923077}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={75.92307692307692}
          x={307.6923076923077}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={75.92307692307692}
          x={384.6153846153846}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={75.92307692307692}
          x={461.53846153846155}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={75.92307692307692}
          x={538.4615384615385}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={75.92307692307692}
          x={615.3846153846154}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={75.92307692307692}
          x={692.3076923076923}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={75.92307692307692}
          x={769.2307692307692}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={75.92307692307692}
          x={846.1538461538461}
          y="5"
        />
        <rect
          fill="#FF6666"
          height="10"
          width={75.92307692307692}
          x={923.0769230769231}
          y="5"
        />
      </g>
    </svg>
  </div>
  <div
    style={
      Object {
        "height": 48,
      }
    }
  >
    <span
      className="jobName"
      style={
        Object {
          "position": "absolute",
          "transform": "translateX(76.92307692307692px)",
        }
      }
    >
      Senior
       
      Software Engineer
    </span>
  </div>
</div>
`;
